ipv6/prefix.go

```diff
@@ -0,0 +1,208 @@
+		length: uint32(mask.Length()),
+	}
+}
+
+// parseNet returns a single *net.IPNet that unifies the IP address and
+// the mask. It leaves out the network address which net.ParseCIDR returns.
+// This may be considered an abuse of the IPNet construct as it is documented
+// that IP is supposed to be the "network number". However, the public IPNet
+// interface does not dissallow it and this usage has been spotted in the wild.
+func parseNet(prefix string) (*net.IPNet, error) {
+	ip, ipNet, err := net.ParseCIDR(prefix)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse prefix: %s", prefix)
+	}
+	return &net.IPNet{IP: ip, Mask: ipNet.Mask}, nil
+}
+
+// ParsePrefix returns the net represented by `addr` in dotted-quad CIDR
+// notation. If the prefix cannot be parsed, then error is non-nil and the
+// prefix returned must be ignored.
+func ParsePrefix(prefix string) (Prefix, error) {
+	ipNet, err := parseNet(prefix)
+	if err != nil {
+		return Prefix{}, err
+	}
+	return PrefixFromNetIPNet(ipNet)
+}
+
+// Address returns the address part of the Prefix, including host bits
+func (me Prefix) Address() Address {
+	return me.addr
+}
+
+// Prefix implements PrefixI
+func (me Prefix) Prefix() Prefix {
+	return me
+}
+
+// ToNetIPNet returns a *net.IPNet representation of this prefix
+func (me Prefix) ToNetIPNet() *net.IPNet {
+	return &net.IPNet{
+		IP:   me.addr.ToNetIP(),
+		Mask: net.CIDRMask(me.Length(), addressSize),
+	}
+}
+
+// lessThan reports whether this Prefix comes strictly before `other`
+// lexigraphically.
+func (me Prefix) lessThan(other Prefix) bool {
+	meNet, otherNet := me.Network(), other.Network()
+	if meNet.addr != otherNet.addr {
+		return meNet.addr.lessThan(otherNet.addr)
+	}
+	if me.length == other.length {
+		return me.Host().addr.lessThan(other.Host().addr)
+	}
+	return me.length < other.length
+}
+
+// Length returns the number of leading 1s in the mask.
+func (me Prefix) Length() int {
+	return int(me.length)
+}
+
+// Mask returns a new Address with 1s in the first `length` bits and then 0s
+// representing the network mask for this prefix.
+func (me Prefix) Mask() Mask {
+	return lengthToMask(me.Length())
+}
+
+// Network returns a new Prefix with all bits after `length` zeroed out so that
+// only the bits in the `network` part of the prefix are present. Note that
+// this method ignores special cases where a network address doesn't make
+// sense like in a host route or point-to-point prefix (/128 and /127). It just
+// does the math.
+func (me Prefix) Network() Prefix {
+	network := me.addr.ui.and(me.Mask().ui)
+	return Prefix{
+		addr: Address{
+			ui: network,
+		},
+		length: me.length,
+	}
+}
+
+// Broadcast returns a new Prefix with all bits after `length` set to 1s. Note
+// that this method ignores special cases where a broadcast address doesn't
+// make sense like in a host route or point-to-point prefix (/128 and /127). It
+// just does the math.
+func (me Prefix) Broadcast() Prefix {
```

Comment:
There is no such thing as a broadcast address in IPv6. This method should not be here.